File: java/src/main/java/S034.java
import com.google.common.math.IntMath;

import java.util.stream.IntStream;

public class S034 {

    // greater than 9! * 7
    private static final int LIMIT = 3 * 1000 * 1000;

    public static int solve() {
        return IntStream.range(10, LIMIT).filter(S034::isCurious).sum();
    }

    private static boolean isCurious(int n) {
        int remaining = n;
        int factorialSum = 0;
        while (remaining > 0) {
            factorialSum += IntMath.factorial(remaining % 10);
            remaining /= 10;
        }
        return factorialSum == n;
    }
}

File: cpp/S034.cpp
#include <iostream>
#include "maths.cpp"

using namespace std;

// 9! * 7
static const int LIMIT = 2540160;

inline bool is_curious(int n) {
    int fac_sum = 0;
    int remaining = n;
    while (remaining > 0) {
        fac_sum += factorial(remaining % 10);
        remaining /= 10;
    }
    return fac_sum == n;
}

int main() {
    int sum = 0;
    for (int i = 10; i < LIMIT; i++) {
        sum += i * is_curious(i);
    }
    cout << sum << endl;
    return 0;
}